src/main/java/edu/umich/its/cpm/Utils.java

```diff
@@ -433,40 +433,93 @@ public static boolean isCurrentUserCPMAdmin(String userId, Environment env) {
 		String testUser = request.getParameter(TEST_USER);
-		boolean isTestUrlEnabled = Boolean.valueOf(env.getProperty(ALLOW_USER_URLOVERRIDE));
+		boolean allowTestUserOverride = ((env.getProperty(ALLOW_USER_URLOVERRIDE) == null ? false : Boolean.valueOf(env.getProperty(ALLOW_USER_URLOVERRIDE))));
 		String testUserInSession = (String) request.getSession().getAttribute(TEST_USER);
 
-		if (isTestUrlEnabled && testUser != null) {
-			user = testUser;
-			request.getSession().setAttribute(TEST_USER, testUser);
-			log.debug("TEST USER is " + user);
-			return user;
+		// simple authenticated user
+		if (!allowTestUserOverride) {
+			return remoteUser;
 		}
-		if (isTestUrlEnabled && testUserInSession != null) {
-			user = testUserInSession;
-			log.debug("TEST USER FROM SESSION " + user);
-			return user;
+
+		if (allowTestUserOverride) {
+
+			// user is authenticated, can become user using "testUser" attribute so we authorize the user using LDAP.
+			// only admin's can become users
+			if (remoteUser != null) {
+				//become user case
+				if (testUser != null) {
+					if (isCurrentUserCPMAdmin(remoteUser, env) && (!testUser.contentEquals(remoteUser))) {
+						user = testUser;
+						request.getSession().setAttribute(TEST_USER, testUser);
+						log.info("BECOME USER " + user);
+					} else {
+						user = makeUserRemoteUserAndRemoveTestUserFromSession(request, remoteUser);
+
+					}
+					return user;
+				}
+				//become user case
+				if (testUserInSession != null) {
+					//getting the user from session as subsequent request from UI won't provide the
+					// testUser attribute as part of request. checking for each call for authorization, safe to have this in place.
+					if (isCurrentUserCPMAdmin(remoteUser, env)) {
+						user = testUserInSession;
+						log.info("BECOME USER IN SESSION: " + user);
+					} else {
+						user = makeUserRemoteUserAndRemoveTestUserFromSession(request, remoteUser);
+					}
+
+					return user;
+				}
+				// authenticated user
+				user = remoteUser;
+				log.info("REMOTE USER " + user);
+				return user;
+			}
+			// no authentication getting the user from testUser attribute
+			if(remoteUser==null) {
+				if (testUser != null) {
+					user = testUser;
+					request.getSession().setAttribute(TEST_USER, testUser);
+					log.info("TEST USER is " + user);
+					return user;
+				}
+				// no authentication case
+				if (testUserInSession != null) {
+					user = testUserInSession;
+					log.info("TEST USER FROM SESSION " + user);
+					return user;
+				}
+			}
+
+			if (user == null) {
+				log.error("No proper user could be retrieved from request");
+			}
 		}
+		return user;
+	}
 
-		if (!isTestUrlEnabled && remoteUser != null) {
+		// this is a case where a non admin user try's to become user or a admin try to use testUser
+		//feature to become then self.
+		private static String makeUserRemoteUserAndRemoveTestUserFromSession(HttpServletRequest request, String remoteUser) {
```

Comment:
Indentation.

Why is it necessary to remove the attribute? Won't it still work if the user in session is the original user?  

Maybe it works to always think of the session user as the current user and only use the remote user to:
- default the session user to the remote user if the session user is set yet.
- verify that the original, logged in (remote), user has authority to actually use the testUser parameter to change the session user.